### PR TITLE
Github action to compute Dora metrics for a service

### PR DESCRIPTION
**Pull request by mk-armah**
Github action computes dora metrics for a service (repo)

## Inputs
| Name                 | Description                                                                                          | Required | Default            |
|----------------------|------------------------------------------------------------------------------------------------------|----------|--------------------|
| repository              | GitHub repository in owner/repo format eg. port-labs/self-service-actions                               | true     | -                  |
| timeframe              | Time frame in weeks to calculate metrics on                                                              | true    | -               |

### CODE REVIEW

**Comment by awconstable** [comment]
Look forward to using this

File: github-workflows/github/src/calculate_pr_metrics.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import json
+
+class RepositoryMetrics:
+    def __init__(self, repo_name, time_frame):
+        self.g = Github(os.getenv('GITHUB_TOKEN'))
+        self.repo_name = repo_name
+        self.time_frame = int(time_frame)
+        self.start_date = datetime.utcnow().replace(tzinfo=timezone.utc) - timedelta(days=self.time_frame)
+        self.repo = self.g.get_repo(f"{self.repo_name}")
+
```
Comment (Tankilevitch) [inline github-workflows/github/src/calculate_pr_metrics.py:14]: overall in python I prefer that we use more descriptive parameter names.
```suggestion
        self.github_client = Github(os.getenv('GITHUB_TOKEN'))
        self.repo_name = repo_name
        self.time_frame = int(time_frame)
        self.start_date = datetime.utcnow().replace(tzinfo=timezone.utc) - timedelta(days=self.time_frame)
        self.repo = self.g.get_repo(f"{self.repo_name}")


```

File: github-workflows/github/dora-metrics.yaml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.x'
+      
+      - name: Install Python dependencies
+        run: |
+          python -m pip install --upgrade pip
+          # pip install httpx
+          pip install PyGithub datetime
```
Comment (Tankilevitch) [inline github-workflows/github/dora-metrics.yaml:100]: create `requirements.txt` containing those versions and install it

File: github-workflows/github/README.md [hunk 3]
```diff
@@ -0,0 +1,401 @@
+<img align="right" width="100" height="74" src="https://user-images.githubusercontent.com/8277210/183290025-d7b24277-dfb4-4ce1-bece-7fe0ecd5efd4.svg" />
+
+# Dora Metrics 
+
+[![Slack](https://img.shields.io/badge/Slack-4A154B?style=for-the-badge&logo=slack&logoColor=white)](https://join.slack.com/t/devex-community/shared_invite/zt-1bmf5621e-GGfuJdMPK2D8UN58qL4E_g)
```
Comment (Tankilevitch) [inline github-workflows/github/README.md:5]: update slack to port-community slack link

File: github-workflows/github/dora-metrics.yaml [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+          logMessage: "Computing lead time for changes metric ..."
+          
+      - name: Run DORA Lead Time for Changes
+        id: lead_time_for_changes
+        shell: pwsh
+        run: |
+          $scriptPath = Join-Path ${{ github.workspace }} "src/leadtimeforchanges.ps1"
+          chmod +x $scriptPath
+          $output = & pwsh -File $scriptPath -ownerRepo "${{ github.event.inputs.repository }}" -workflows "CI/CD" -branch "main" -numberOfDays '30' -commitCountingMethod "last" -patToken "${{ secrets.PATTOKEN }}"
+          $jsonLine = $output | Select-String "###JSON_START###" -Context 0,1 | ForEach-Object { $_.Context.PostContext }
```
Comment (Tankilevitch) [inline github-workflows/github/dora-metrics.yaml:78]: I see that you are using both `secrets.PATTOKEN` and `secret.GITHUB_TOKEN` please explain why do we need both?

File: github-workflows/github/dora-metrics.yaml [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+          clientSecret: ${{ secrets.PORT_CLIENT_SECRET }}
+          baseUrl: https://api.getport.io
+          operation: PATCH_RUN
+          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
+          logMessage: "Upserting dora metrics to Port"
+
+      - name: UPSERT Entity
+        uses: port-labs/port-github-action@v1
+        with:
+          identifier: "${{ fromJson(env.metrics).id }}"
```
Comment (Tankilevitch) [inline github-workflows/github/dora-metrics.yaml:131]: where is the `env.metrics` is coming from? couldn't find it

File: github-workflows/github/dora-metrics.yaml [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        description: 'Comma-separated list of repositories to analyze (eg. port-labs/self-service-actions)'
+        required: true
+      timeframe:
+        description: 'Last X weeks'
+        required: true
+      port_payload:
+        required: true
+        description: Port's payload, including details for who triggered the action and
+          general context (blueprint, run id, etc...)
+        type: string
```
Comment (Tankilevitch) [inline github-workflows/github/dora-metrics.yaml:16]: I think it is possible that the action will run scheduled in their environment and not from an action, make sure this is supported.

File: github-workflows/github/src/calculate_pr_metrics.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        }
+
+        return metrics
+
+    def timedelta_to_decimal_hours(self, td):
+        return round(td.total_seconds() / 3600, 2)
+
+def main():
+    repo_name = os.getenv('REPOSITORY')
+    time_frame = os.getenv('TIMEFRAME_IN_DAYS') #os.getenv('TIME_FRAME')
```
Comment (Tankilevitch) [inline github-workflows/github/src/calculate_pr_metrics.py:107]:
```suggestion
    time_frame = os.getenv('TIMEFRAME_IN_DAYS')

```

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: github-workflows/github/README.md [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+
+```json
+{
+  "identifier": "dora_metrics",
+  "title": "Dora Metrics",
+  "icon": "Github",
+  "userInputs": {
+    "properties": {
+      "timeframe": {
+        "icon": "Github",a
```
Comment (PeyGis) [inline github-workflows/github/README.md:63]: There's a typo here

File: github-workflows/github/README.md [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+      "timeframe",
+      "repository"
+    ],
+    "order": [
+      "timeframe"
+    ]
+  },
+  "invocationMethod": {
+    "type": "GITHUB",
+    "org": "my-org",
```
Comment (PeyGis) [inline github-workflows/github/README.md:85]: update to `<your-org>`

File: github-workflows/github/README.md [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+    "repo": "my-org",
```
Comment (PeyGis) [inline github-workflows/github/README.md:86]: update to `<your-repo>`

**Review by PeyGis — CHANGES_REQUESTED** [review]

File: github-workflows/github/dora-metrics.yaml (hunk 5 above)
Comment (mk-armah) [reply github-workflows/github/dora-metrics.yaml:131]: Python Script

**Comment by awconstable** [comment]
Please can you explain why this was closed? Is there an alternative approach?

**Pull request closed without merging**